File: stack/155_min_stack/Solution.java
class MinStack {
    public List<int[]> stack ;
    public MinStack() {
        this.stack =  new ArrayList<>();
    }
    
    public void push(int val) {
       
        int min = stack.isEmpty() ? val : Math.min(val, getMin());
        this.stack.add(new int[]{val,min});
    }
    
    public void pop() {
        this.stack.remove(this.stack.size()-1);
    }
    
    public int top() {
        return this.stack.get(this.stack.size()-1)[0];
    }
    
    public int getMin() {
        return this.stack.get(this.stack.size()-1)[1];
    }
}

/**
 * Your MinStack object will be instantiated and called as such:
 * MinStack obj = new MinStack();
 * obj.push(val);
 * obj.pop();
 * int param_3 = obj.top();
 * int param_4 = obj.getMin();
 */

File: stack/155_min_stack/Solution.py
class MinStack:

    def __init__(self):
        self.stack = []

    def push(self, val: int) -> None:
        if len(self.stack) == 0:
            temp = [val,val]
            self.stack.append(temp)
        else:
            min_val = min(self.stack[-1][1],val)
            temp =[val,min_val]
            self.stack.append(temp)

    def pop(self) -> None:
        return self.stack.pop()[0]

    def top(self) -> int:
        return self.stack[-1][0]

    def getMin(self) -> int:
        return self.stack[-1][1]
        


# Your MinStack object will be instantiated and called as such:
# obj = MinStack()
# obj.push(val)
# obj.pop()
# param_3 = obj.top()
# param_4 = obj.getMin()